cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+		}
+		allowedReplicasets = append(allowedReplicasets, replicaset)
+	}
+
+	return allowedReplicasets, nil
+}
+
+func Upgrade(replicasets replicaset.Replicasets, pendingReplicasetAliases []string,
+	lsnTimeout int) error {
+	allowedReplicasets, err := GetAllowedReplicasets(replicasets, pendingReplicasetAliases)
+	if err != nil {
+		return err
+	}
+
+	printReplicasetStatus := func(alias, status string) {
+		fmt.Printf("• %s: %s\n", alias, status)
+	}
+
+	for _, rs := range allowedReplicasets {
+		var masterRun *running.InstanceCtx = nil
+		var replicRun []running.InstanceCtx
+
+		for _, inst := range rs.Instances {
+			run := inst.InstanceCtx
+			fullInstanceName := running.GetAppInstanceName(run)
+			conn, err := connector.Connect(connector.ConnectOpts{
+				Network: "unix",
+				Address: run.ConsoleSocket,
+			})
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+
+			res, err := conn.Eval(
+				"return (type(box.cfg) == 'function') or box.info.ro",
+				[]any{}, connector.RequestOpts{})
+			if err != nil || len(res) == 0 {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+
+			if !res[0].(bool) {
+				if masterRun != nil {
+					printReplicasetStatus(rs.Alias, "error")
+					return fmt.Errorf("[%s]: %s and %s are both masters",
+						rs.Alias, running.GetAppInstanceName(*masterRun),
+						fullInstanceName)
+				}
+				masterRun = &run
+			} else {
+				replicRun = append(replicRun, run)
+			}
+		}
+		if masterRun == nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s]: has not master instance", rs.Alias)
+		}
+		var conn connector.Connector
+		conn, err = connector.Connect(connector.ConnectOpts{
+			Network: "unix",
+			Address: masterRun.ConsoleSocket,
+		})
+
+		if err != nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s][%s]: %s", rs.Alias,
+				running.GetAppInstanceName(*masterRun), err)
+		}
+		masterUpgradeInfo, err := internalUpgrade(conn)
+		if err != nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s][%s]: %s", rs.Alias,
+				running.GetAppInstanceName(*masterRun), err)
+		}
+		for _, run := range replicRun {
+			fullInstanceName := running.GetAppInstanceName(run)
+			conn, err = connector.Connect(connector.ConnectOpts{
+				Network: "unix",
+				Address: run.ConsoleSocket,
+			})
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+			err = WaitLSN(conn, masterUpgradeInfo, lsnTimeout)
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s]: LSN wait timeout: error waiting LSN %d "+
+					"in vclock component %d on %s: time quota %d seconds "+
+					"exceeded", rs.Alias, masterUpgradeInfo.LSN,
+					masterUpgradeInfo.IID, fullInstanceName, lsnTimeout)
+			}
```

Comment:
@psergee Let's discuss it in a thread (to ease navigation over the discussion later).

> > ⨯ [storage-001]: LSN wait timeout: error waiting LSN 2003085 in vclock component 1
>
> This message about LSN seems a bit confusing to me. Maybe add something like "Schema upgrade replication timeout exceeded:" before "error waiting LSN..." to make it clearer what has just happened?

I agree that it may be convenient to have some reason, why the LSN waiting is needed in the first place. However, 'schema upgrade replication timeout' seems a bit vague for me. How about the following variant?

> ⨯ [storage-001-replica]: can't ensure that upgrade operations performed on "storage-001-master" are replicated to "storage-001-replica" to perform snapshotting on it: error waiting LSN 2003085 in vclock component 1: <..last error from eval or actual LSN..>

Reply:
I chose the last from the ones suggested.

Reply:
Thanks!

@psergee Is it OK for you?